### PR TITLE
Support linking Vulkan directly

### PR DESCRIPTION
**Pull request by Ralith**
This is the preferred pattern in most environments when an application cannot function without Vulkan, as it saves the libloading dependency, eliminates an error case, and makes the Vulkan dependency visible to the OS.

Ideally I'd like to move towards this being the default, but that might be more disruptive than is justified.

Draft until I've tested this properly on Windows. The build script might need to explicitly add `VULKAN_SDK/lib` to the linker search path.

### CODE REVIEW

**Review by MarijnS95 — COMMENTED** [review]
Nice, I've been intending to [propose the same](https://github.com/EmbarkStudios/ash-molten/pull/56#issuecomment-877034275) after `0.33` is out assuming we can get permission to absorb detection and static linking of `MoltenVK` through the same mechanism, effectively making `ash-molten` obsolete.

I'm still conflicted on the `linked` name, but guess it makes sense if we were to support both dynamic _and_ static linking at some point in time.

File: ash/src/entry_linked.rs
Comment (MarijnS95) [reply ash/src/entry_linked.rs:12]: Unfortunately this is still "broken", see https://github.com/rust-lang/rust/issues/83428:

image

Feature requirements are propagated to the functions _within_ types in this module, but not to the type itself. Turn this into `pub mod` and it all starts working.

**Comment by Friz64** [comment]
Please be aware that the current implementation (indirect linking) is recommended in order to achieve optimal performance as per the [Vulkan Loader documentation](https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#best-application-performance-setup).

**Comment by MarijnS95** [comment]
@Friz64 You probably intended to link https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#indirectly-linking-to-the-loader as https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#best-application-performance-setup is specific to linking Vulkan functions **after** having acquired `vkGetInstanceProcAddr`/`vkGetDeviceProcAddr`. This PR only concerns itself with dynamic/runtime loading of the `vkGetInstanceProcAddr` entrypoint, which should theoretically be faster if performed by the library loader instead of having to go through extra hoops in userspace to call `dlopen` + `dlsym` (though this is what `ld.so` performs under the hood...).

**Comment by Friz64** [comment]
Yes, i intended to link that one instead. The documentation is somewhat confusing in this regard, but i understand now.

File: ash/src/entry_libloading.rs
Comment (Friz64) [reply ash/src/entry_libloading.rs:58]: This is my biggest concern as the Vulkan library will need to be present during linking and thus break CIs. One could indeed install the Vulkan Loader on the CI server somehow but i think that's too much of a hassle.

Comment (MarijnS95) [reply ash/src/entry_libloading.rs:58]: @Friz64 I'm curious if this isn't the case already. GH Actions - which a lot of Rust crates use - is already loaded to the brim with packages.

Comment (Friz64) [reply ash/src/entry_libloading.rs:58]: Good thought, but [does not look like that is the case, unfortunately](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md).

File: ash/src/entry_linked.rs
Comment (Ralith) [reply ash/src/entry_linked.rs:12]: So, no action needed for now then?

File: ash/src/entry_linked.rs
Comment (Ralith) [reply ash/src/entry_linked.rs:8]: Do you have a specific proposal for a better wording? I'd like to have something reasonably concise that still gives a sense of the advantages.

File: ash/src/entry_libloading.rs
Comment (Ralith) [reply ash/src/entry_libloading.rs:58]: > If all these questions are answered with no we should probably seek to move in this direction by default, instead of using libloading by default.

I agree. The only reason I'm aware of to use dlopen for Vulkan is if you have a graceful fallback path, which is an extremely uncommon case, well suited to requiring an opt-in.

> One could indeed install the Vulkan Loader on the CI server somehow but i think that's too much of a hassle.

If this is an issue, fixing it is as simple as `apt-get install libvulkan1` or whatever--trivial. But CI seems to be passing fine with `--all-features` already. The loader is part of the base install of most distros these days, after all.

File: ash/src/entry_libloading.rs
Comment (Ralith) [reply ash/src/entry_libloading.rs:59]: IMO this is a must for feature-gated APIs; it's this or less-visible handwritten warnings. Unless the automagic version is imminent we should do whatever's necessary to have good docs.

File: ash/src/entry.rs
Comment (Ralith) [reply ash/src/entry.rs:31]: `load` is the relevant interface for this particular function; `StaticFn::get_instance_proc_addr` is not directly exposed, but fetched internally via `load`, which might return garbage.

File: ash/src/entry_linked.rs
Comment (MarijnS95) [reply ash/src/entry_linked.rs:8]: `Function loader from dynamically linked entrypoint` (`s/dynamically linked/runtime-loaded` for `Entry`?

File: ash/src/entry_libloading.rs
Comment (MarijnS95) [reply ash/src/entry_libloading.rs:58]: > I agree. The only reason I'm aware of to use dlopen for Vulkan is if you have a graceful fallback path, which is an extremely uncommon case, well suited to requiring an opt-in.

Exactly this :+1: 
 
> > One could indeed install the Vulkan Loader on the CI server somehow but i think that's too much of a hassle.
> 
> If this is an issue, fixing it is as simple as `apt-get install libvulkan1` or whatever--trivial. But CI seems to be passing fine with `--all-features` already. The loader is part of the base install of most distros these days, after all.

I couldn't find it in the same page @Friz64 linked, and `--all-features` on `check` does not perform the final linking step where the library is needed, we should use `cargo build` for that. That's probably a good thing, and for projects needing to spit out built artifacts they can easily add it with `apt` - that's the justification for [GH actions not including `Vulkan` (the SDK) in the first place](https://github.com/actions/virtual-environments/issues/18).

File: ash/src/entry_linked.rs
Comment (MarijnS95) [reply ash/src/entry_linked.rs:12]: Nope, it's unfortunate but it is what it is. I hope to notice the GH link whenever that issue is closed and move it at some point.

**Comment by Ralith** [comment]
> I'm still conflicted on the linked name, but guess it makes sense if we were to support both dynamic and static linking at some point in time.

None of this code is sensitive to whether the linked library is static or dynamic; it's just a matter of what's in your search path.

File: ash/src/entry_libloading.rs
Comment (MarijnS95) [reply ash/src/entry_libloading.rs:59]: Let's keep it for now, it's better than nothing at all. Glad docs.rs uses nightly to build at least.

File: ash/src/entry_libloading.rs
Comment (Ralith) [reply ash/src/entry_libloading.rs:60]: If we make linking the loader the default I'm in favor of this. For simplicity, we might even consider something [like I'm doing for OpenXR](https://github.com/Ralith/openxrs/blob/master/openxr/src/entry.rs#L262-L266), where there's a single monomorphic `Entry` type that has a cfg-guarded library guard.

**Comment by MarijnS95** [comment]
> > I'm still conflicted on the linked name, but guess it makes sense if we were to support both dynamic and static linking at some point in time.
> 
> None of this code is sensitive to whether the linked library is static or dynamic; it's just a matter of what's in your search path.

It feels like there should be something to distinguish it from runtime linking/loading, perhaps renaming `libloading` because that's rather vague if unfamiliar with the particular crate name (and anyway the only feature - and selected by default - so I don't think anyone has that explicitly listed in their `Cargo.toml`).

**Comment by Ralith** [comment]
`dlopen` isn't any kind of linking at all, so I think `linked` is a pretty apt distinction; renaming the other feature makes sense, and shouldn't be very disruptive since it's defaulted on.

File: ash/src/entry_libloading.rs
Comment (Ralith) [reply ash/src/entry_libloading.rs]: Switched to linking by default.

File: ash/src/entry.rs
Comment (Ralith) [reply ash/src/entry.rs:37]: Added safety docs.

File: ash/src/entry_libloading.rs
Comment (Ralith) [reply ash/src/entry_libloading.rs]: Shuffled the names around.

File: ash/src/entry_linked.rs
Comment (Ralith) [reply ash/src/entry_linked.rs]: Updated in that spirit, thanks!

**Comment by Ralith** [comment]
I've added a commit which folds all the various entry types into a single monomorphic `Entry` struct that contains a lib guard iff runtime loading is enabled, just like I do in the OpenXR bindings. I think this is a big win for API simplicity.

**Comment by Ralith** [comment]
Findings from CI: libvulkan.so.1 (which is what ends up in the binary's RUNPATH) *is* installed by default, but libvulkan.so (which is the symlink the linker finds at compiletime) comes from libvulkan-dev, which is not.

**Comment by Rua** [comment]
The `Entry` API currently assumes that you will be using the "fancy" wrappers, but Vulkano uses only the bare wrappers, `EntryFnV1_0` and so on. Therefore, Vulkano has kept using its original loader for the moment. Would it be possible to change the loading mechanism in such a way that it can be used with the bare wrappers as well? Then Vulkano could use it. Thanks!

**Comment by Ralith** [comment]
What exactly about the current iteration poses a problem for you? You've always been able to access the function pointer table directly via e.g. `Entry::fp_v1_0`.

**Comment by Rua** [comment]
True, but it feels a bit weird to construct an `Entry` when I don't intend to use it at all. Separating the loading mechanism from the rest of the Vulkan API seems cleaner.

**Comment by Ralith** [comment]
How exactly do you not intend to use it? `Entry` loads function pointers, which is what you're asking for access to.

File: examples/src/lib.rs
Comment (MarijnS95) [inline examples/src/lib.rs:10]: That's a strange change.

File: ash/src/lib.rs
Comment (MarijnS95) [inline ash/src/lib.rs:32]:
```suggestion
//! Load the Vulkan library linked at compile time using [`Entry::new()`], or at load it at runtime
//! using [`Entry::load()`], which uses `libloading`.
```

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:136]: I like this better than the previous implementation forcibly loading `b"vkGetInstanceProcAddr\0"` through the custom loader function; this truly allows a custom function to be passed without hackery like https://github.com/EmbarkStudios/ash-molten/blob/5e30540b75954160c5a62d951a65f0c9816fff15/src/lib.rs#L91-L97.

(But besides manually constructing `vk::StaticFn` an application is still completely free to call `vk::StaticFn::load(_checked)` :tada:)

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:32]: Maybe

```suggestion
    /// Load entrypoint from a linked Vulkan loader
```

?

This could help clarify the `Note` below.

File: ash/src/lib.rs
Comment (MarijnS95) [inline ash/src/lib.rs:33]: Perhaps leave this last sentence in place, referring to ``[`Entry::from_static_fn()`]`` instead?

File: .github/workflows/ci.yml
Comment (MarijnS95) [inline .github/workflows/ci.yml:19]: While the features can technically be used in parallel, should we `check` the possible combinations in isolation? (ie `linked`, `loaded`, and `linked,loaded`)

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:68]:
```suggestion
    /// `dlopen`ing native libraries is inherently unsafe. The safety guidelines
    /// for [`Library::new()`] and [`Library::get()`] apply here.
```

(I forgot those initially)

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:110]:
```suggestion
    /// for [`Library::new()`] and [`Library::get()`] apply here.
```

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:208]: :+1:

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:133]: Nit: None of the `# Safety` markdown headers have an empty line after them (this anyway isn't visible in rendered documentation).

```suggestion
    /// # Safety
```

**Comment by MarijnS95** [comment]
> `dlopen` isn't any kind of linking at all, so I think `linked` is a pretty apt distinction; renaming the other feature makes sense, and indeed shouldn't be very disruptive.

While I would technically agree, this isn't the first time I have seen it being called a [`dynamic linking loader`](https://linux.die.net/man/3/dlopen).

> I've added a commit which folds all the various entry types into a single monomorphic `Entry` struct that contains a lib guard iff runtime loading is enabled, just like I do in the OpenXR bindings. I think this is a big win for API simplicity.

Should be fine, it's extremely unlikely for the features to be used together. In the rare event that someone wishes to use multiple Vulkan libraries at the same time (building whatever arbitration and circumventing the ICD) they better use runtime loading to keep the libraries separated.

> Findings from CI: libvulkan.so.1 (which is what ends up in the binary's RUNPATH) _is_ installed by default, but libvulkan.so (which is the symlink the linker finds at compiletime) comes from libvulkan-dev, which is not.

Fortunately we're already using `libvulkan.so.1` in `entry_libloading.rs`, not `libvulkan.so`, but that isn't easily selectable with `println!("cargo:rustc-link-lib=vulkan");` I suppose? Pretty strange of Ubuntu to provide it that way...

> The `Entry` API currently assumes that you will be using the "fancy" wrappers, but Vulkano uses only the bare wrappers, `EntryFnV1_0` and so on. Therefore, Vulkano has kept using its original loader for the moment. Would it be possible to change the loading mechanism in such a way that it can be used with the bare wrappers as well? Then Vulkano could use it. Thanks!

@Ralith It appears @Rua is referring to this shorthand:

https://github.com/MaikKlein/ash/blob/17149bd791cd6e09b145063a238d3e27d855780c/ash/src/entry.rs#L159-L166

Which allows all structs in `ash/src/extensions` to accept an `Entry` to call `get_instance_proc_addr()` on, instead of the raw `vk::StaticFn`. I suppose Vulkano is keeping track of the various `vk::*Fn*` structs on its own?

But that has always been the case. Not sure if it should change towards `StaticFn` as part of this PR, the lines are already being touched though. In the end I'm in favour of this change as Ash should just accept the bare minimum (`StaticFn` in this case) in its API.

**Review by MarijnS95 — COMMENTED** [review]
This is looking really good so far. Should we let the CI run on it? I'm curious what others think on it, @MaikKlein? It's quite an impactful change, but I'd love to roll this out soon.

File: .github/workflows/ci.yml
Comment (Ralith) [reply .github/workflows/ci.yml:19]: They don't really interact, and in general testing every combination of features is combinatorically infeasible, so my inclination is not to bother.

**Review by Ralith — COMMENTED** [review]
> Should we let the CI run on it?

CI seems to be running on it fine here? Maybe that's only visible to me. Will un-draft it once I've satisfied my concerns about the Windows build.

> @Ralith It appears @Rua is referring to this shorthand:

I don't really follow. What capability is missing from the refactored `Entry`?

> Pretty strange of Ubuntu to provide [libvulkan.so in the -dev package]

I think this is reasonable for them to do, as the unversioned .so symlink is only referenced at linktime. If we somehow hardcoded the .so.1 in the linker args that would just make us fragile and unidiomatic.

> it's extremely unlikely for the features to be used together

To be clear, they can still be used together with the unified type; cargo features must be composable, after all.

> While I would technically agree, this isn't the first time I have seen it being called a dynamic linking loader.

Hah, good cite. I'm open to clearer naming ideas.

**Comment by Ralith** [comment]
Added some magic to build.rs to make this Just Work with a Vulkan SDK install on Windows (and Linux, if for some reason you aren't using the proper distro packages).

**Comment by Ralith** [comment]
If anyone has any ideas what, if any, special measures are needed for good behavior on macOS, let me know!

File: ash/build.rs
Comment (MaikKlein) [inline ash/build.rs:22]: Can you remind me what kind of linking this is? Will this always be static linking?

Comment (MaikKlein) [reply ash/build.rs:22]: fwiw only having static linking is fine I think, I don't think we need dynamic linking because we have runtime loading which are fairly similar.

**Review by MaikKlein — COMMENTED** [review]
Awesome, I don't have any concerns here.

> If anyone has any ideas what, if any, special measures are needed for good behavior on macOS, let me know!

Not sure myself. I am also not sure how this will integrate with [ash-molten](https://github.com/EmbarkStudios/ash-molten/).

I'd like to keep ash-molten as a separate crate because it is quite opinionated an builds or downloads moltenvk itself.

I think we can keep ash-molten as is and it can internally do `Entry::from_static_fn`. I think we can use no-default-features there as well and let ash-molten statically link with vulkan.

File: ash/build.rs
Comment (Ralith) [reply ash/build.rs:22]: Per [rustc docs](https://doc.rust-lang.org/rustc/command-line-arguments.html#option-l-link-lib), this will link whatever type of library is available, favoring dynamic if both are found. Note that the Vulkan loader is normally distributed only as a dynamic library.

> I think we can keep ash-molten as is and it can internally do `Entry::from_static_fn`. I think we can use no-default-features there as well and let ash-molten statically link with vulkan.

This should work great.

File: .github/workflows/ci.yml
Comment (MarijnS95) [reply .github/workflows/ci.yml:19]: There are only two features that are intended to be used separately, but fine to leave as-is.

**Comment by MarijnS95** [comment]
> > Should we let the CI run on it?
> 
> CI seems to be running on it fine here? Maybe that's only visible to me. Will un-draft it once I've satisfied my concerns about the Windows build.

Yeah, that. Good to see it run now.

> > @Ralith It appears @Rua is referring to this shorthand:
> 
> I don't really follow. What capability is missing from the refactored `Entry`?

Nothing that this PR changes, afaik. They must not be using `Entry` at all for whatever reason. The only change I can see being useful is to require just `StaticFn` in extension constructors instead of the "full" `Entry` which is not technically required to call `StaticFn::get_instance_proc_addr`.

> > Pretty strange of Ubuntu to provide [libvulkan.so in the -dev package]
> 
> I think this is reasonable for them to do, as the unversioned .so symlink is only referenced at linktime. If we somehow hardcoded the .so.1 in the linker args that would just make us fragile and unidiomatic.

Derp. Yes, that is totally sensible. We're already using `.so.1` for linker args though, right? Looking at `ldd $(which vkcube)` on my machine (Arch Linux) shows links against the versioned variant too. Seems like the most sane thing to do compatibility-wise.

> > it's extremely unlikely for the features to be used together
> 
> To be clear, they can still be used together with the unified type; cargo features must be composable, after all.

Exactly. That's better than making them incompatible and breaking `--all-features` in strange ways, thanks!

> > While I would technically agree, this isn't the first time I have seen it being called a dynamic linking loader.
> 
> Hah, good cite. I'm open to clearer naming ideas.

With both features enabled by default I don't think it matters too much anymore. `dynamic_linked` and `runtime_linked` are a bit too verbose.

File: ash/src/entry.rs
Comment (MarijnS95) [inline ash/src/entry.rs:73]: Nit: the rest of the examples have the `?` inside the `unsafe{}` block. Should we move them out in every occasion?

**Review by MarijnS95 — APPROVED** [review]
Ready to get in. Not sure if we should bite the bullet and accept `StaticFn` instead of `Entry` in all the extension constructors though, this PR is already touching all those lines anyway.

**Comment by Ralith** [comment]
> With both features enabled by default

This PR disables dynamic loading by default, since it's a niche case and brings in an extra dependency. I think the docs still do a reasonably good job of making things clear thanks to `doc(cfg(_))` though.

> Not sure if we should bite the bullet and accept StaticFn instead of Entry in all the extension constructors though, this PR is already touching all those lines anyway.

I'd prefer to leave that for future work, since it's logically independent and you can easily construct an `Entry` given a `StaticFn` already.

File: ash/src/entry.rs
Comment (Ralith) [reply ash/src/entry.rs]: This inconsistency predates this PR, but I'll go ahead and clean it up. IMO it's more readable inside the block.

Comment (MarijnS95) [reply ash/src/entry.rs]: Fine by me, it's just an example.

**Comment by MarijnS95** [comment]
Then we'll wait for @MaikKlein to give the final sign-off on this, and get it in! Let's immediately cut a new minor release as well with perhaps some other PRs?

File: ash/src/lib.rs
Comment (filnet) [inline ash/src/lib.rs:31]: typo: "[...] or at load it [...]"

Comment (Ralith) [reply ash/src/lib.rs]: Fixed, thanks!

**Comment by Ralith** [comment]
@MaikKlein ping?

**Comment by MaikKlein** [comment]
Sorry for the delay, looks good :partying_face:  and I'd like to get this in.

Two high level comments:

- This will default to dynamic/static loading which will most likely break all the CIs with a linker error. I assume most don't actually have a Vulkan environment. So I think I'd prefer either runtime loading by default or no default features at all.  With "no default features " at least they would get a build error.

- Does this PR solve https://github.com/MaikKlein/ash/issues/462? How would one build ash statically wit this PR? What happens if you have both the .so and .a in the path? IIRC you can specify the link type somewhere as a flag.  Just asking, this PR doesn't have to solve this.

**Comment by MarijnS95** [comment]
> Sorry for the delay, looks good  and I'd like to get this in.
> 
> Two high level comments:
> 
> * This will default to dynamic/static loading which will most likely break all the CIs with a linker error. I assume most don't actually have a Vulkan environment. So I think I'd prefer either runtime loading by default or no default features at all.  With "no default features " at least they would get a build error.

According to https://github.com/MaikKlein/ash/pull/457#issuecomment-890278798 this needs `sudo apt-get install libvulkan-dev` as seen in `ci.yaml`.

> * Does this PR solve [Support for a statically linked vulkan library #462](https://github.com/MaikKlein/ash/issues/462)? How would one build ash statically wit this PR? What happens if you have both the .so and .a in the path?

According to https://github.com/MaikKlein/ash/pull/457/files#r689123563 the .so is preferred over the .a. I marked #462 as linked to this PR since the asker only seems to have a `.a` (if they're cross-compiling and it finds the hosts' `.so` that's a different issue), probably something to find out over time?

> IIRC you can specify the link type somewhere as a flag.  Just asking, this PR doesn't have to solve this.

There's the optional `[kind=]`, we can add `cargo:rustc-link-lib=static=vulkan` (and `dylib=` just in case) with some feature flag if this turns out to be a problem.

**Comment by MaikKlein** [comment]
> According to #457 (comment) this needs sudo apt-get install libvulkan-dev as seen in ci.yaml.

Yeah that is what I meant. It will break all the CIs without a vulkan environment and they have to add `sudo apt-get install libvulkan-dev` right?

**Comment by Ralith** [comment]
Yes, the build environment will need to have a libvulkan for the target available. We should probably consider this a breaking change for that reason, but IMO it's nonetheless a reasonable default. For one thing, the linker error you get for a missing foreign library is highly recognizable and easy to google, with the same solution across all programming languages, whereas a build error due to not having default features would be more idiosyncratic.

Leaving the default unchanged wouldn't be a disaster, but there's something to be said for defaulting to the lighter-weight, more robust, and more conventional configuration.

> How would one build ash statically wit this PR?

As @MarijnS95 says, it'll use whatever's available in the search path. If the target environment only has a static library, then that's what'll be used. I expect this will be sufficient in practice; especially when cross-compiling, the user has considerable control over the available libraries.

**Comment by Jasper-Bekkers** [comment]
> Yes, the build environment will need to have a libvulkan for the target available. IMO it's a reasonable default. For one thing, the linker error you get for a missing foreign library is highly recognizable and easy to google, with the same solution across all programming languages, whereas a build error due to not having default features would be more idiosyncratic.

I'll resurrect this thread now that ash 0.34 is out and we're starting to see the first linker errors across our crates. I think this may have been a mistake. Rust users aren't used to fixing (or seeing) linker errors, and I honestly think this is a bad precedent to set from a usability point of view.

**Comment by Ralith** [comment]
FWIW, there is overwhelming precedent for this behavior for FFI crates (e.g. cpal, openssl, openexr, gtk, ...). While I'm sure ash 0.34 will be some peoples' first encounter with link-time dependencies, most ash users should find this familiar and easily diagnosed.

**Comment by aloucks** [comment]
I agree with @Jasper-Bekkers, this should not be the default. This imposes a burden that is 100% avoidable.

**Comment by Jasper-Bekkers** [comment]
> I agree with @Jasper-Bekkers. This imposes a burden that is 100% avoidable. This is not the same as a sys crate where we are building the `C` source. We're linking to an external library. I think it's fine to support compile time linking but it should not be the default. 

Along with this, it's quite unclear what the need for this change was to begin with, dropping a small and extremely common dependency is not really a gain. Neither in build times (only if you look at them in isolation, not when you look at them in the context of a much larger app) nor in runtime performance.

At best it's a dubious default to have, at worst it actively harms usability for this crate. I would suggest doing a fox release that reverts this behavior honestly.

**Pull request merged**
